scarb/src/resolver/algorithm/mod.rs

```diff
@@ -1,12 +1,153 @@
 use crate::core::lockfile::Lockfile;
 use crate::core::registry::Registry;
-use crate::core::{Resolve, Summary};
+use crate::core::{PackageId, Resolve, Summary};
+use crate::resolver::algorithm::provider::{
+    rewrite_locked_dependency, DependencyProviderError, PubGrubDependencyProvider, PubGrubPackage,
+};
+use crate::resolver::algorithm::solution::{build_resolve, validate_solution};
+use crate::resolver::algorithm::state::{Request, ResolverState};
+use anyhow::bail;
+use futures::{FutureExt, TryFutureExt};
+use itertools::Itertools;
+use pubgrub::PubGrubError;
+use pubgrub::{DefaultStringReporter, Reporter};
+use pubgrub::{Incompatibility, State};
+use std::collections::HashSet;
+use std::sync::Arc;
+use std::thread;
+use tokio::sync::{mpsc, oneshot};
 
-#[tracing::instrument(level = "trace", skip_all)]
-pub async fn resolve(
-    _summaries: &[Summary],
-    _registry: &dyn Registry,
-    _lockfile: Lockfile,
+mod in_memory_index;
+mod provider;
+mod solution;
+mod state;
+
+pub async fn resolve<'c>(
+    summaries: &[Summary],
+    registry: &dyn Registry,
+    lockfile: Lockfile,
 ) -> anyhow::Result<Resolve> {
-    todo!("implement")
+    let state = Arc::new(ResolverState::default());
+
+    let (request_sink, request_stream): (mpsc::Sender<Request>, mpsc::Receiver<Request>) =
+        mpsc::channel(300);
+
+    let requests_fut = state
+        .clone()
+        .fetch(registry, request_stream)
+        .map_err(|err| anyhow::format_err!(err))
+        .fuse();
+
+    for summary in summaries {
+        for dep in summary.full_dependencies() {
+            let locked_package_id = lockfile.packages_matching(dep.clone());
+            let dep = if let Some(locked_package_id) = locked_package_id {
+                rewrite_locked_dependency(dep.clone(), locked_package_id?)
+            } else {
+                dep.clone()
+            };
+            if state.index.packages().register(dep.clone()) {
+                request_sink.send(Request::Package(dep.clone())).await?;
+            }
+        }
+    }
+
+    let main_package_ids: HashSet<PackageId> =
+        HashSet::from_iter(summaries.iter().map(|sum| sum.package_id));
+
+    let (tx, rx) = oneshot::channel();
+
+    let cloned_lockfile = lockfile.clone();
+    thread::Builder::new()
+        .name("scarb-resolver".into())
+        .spawn(move || {
```

Comment:
consider moving this to a separate named function to: reduce indentation, have something named in stack trace